[feat] 뉴스 기사 목록 조회 기능 구현

yuhandemian opened the pull request:
## Issues
- closed #36 

## ✔️  Check-list
- [x] : Label을 지정해 주세요.
- [x] : Merge할 브랜치를 확인해 주세요.

## 🗒️ Work Description

### 구현 내용
- 기사 목록 조회 API (`GET /api/articles`)를 구현했습니다.
- 기사 단건 조회 API (`GET /api/articles/{articleId}`)를 구현했습니다.
- 출처 목록 조회 API (`GET //api/articles/sources`)를 구현했습니다.

**주요 기능:**
- 커서 기반 페이지네이션
- 다양한 필터링 옵션 (작성자, 관심사, 날짜 범위 등)
- 정렬 기능 (최신순/오래된순)

- **뉴스 기사 상세 조회 API 구현** (`GET /api/articles/{articleId}`)
  - 조회수 자동 증가 및 Redis 기반 24시간 중복 조회 방지
  - `viewedByMe` 필드로 사용자 조회 이력 제공
  - 404/400 예외 처리 추가

- **뉴스 수집 데이터 품질 개선**
  - 한글 콘텐츠 필터링으로 영문 기사 제외
  - 네이버 API 응답 정제 (영문 번역 안내문, URL, 특수기호 제거)
  - 요약 길이 제한 (500자) 및 불필요한 메타데이터 제거

- **예외 처리**:
  - 존재하지 않는 기사: `404 NOT_FOUND` (ArticleNotFoundException)
  - 잘못된 UUID 형식: `400 BAD_REQUEST`

**해결한 문제:**
- 커서 페이지네이션 버그 수정:
  - `nextCursor`가 timestamp 대신 UUID를 반환하도록 수정 (timestamp로 했는데 (안끝나는)무한스크롤 문제 발생했었습니다.)
  - `nextAfter`가 `publishDate` 대신 `createdAt` 기반으로 동작하도록 수정
  - QueryDSL `Tuple` 방식으로 변경하여 `createdAt` 필드 접근 가능

**API 엔드포인트:**

GET /api/articles

## 📷 Screenshot

### Postman
<img width="1261" height="687" alt="image" src="https://github.com/user-attachments/assets/b6828839-5ea2-433c-8076-80d315c686cf" />

### 프론트엔드
<img width="1383" height="598" alt="image" src="https://github.com/user-attachments/assets/164b62f3-1d68-4725-8462-4f049ac2a0e0" />

### 프론트엔드 커서 마지막 페이지 -> "nextCursor": null,"nextAfter": null로 (안끝나는)무한스크롤이 발생하지 않습니다.
<img width="1100" height="643" alt="image" src="https://github.com/user-attachments/assets/a47d62f9-8df8-4f8b-9d4b-6d2efda311b0" />

<img width="1384" height="572" alt="image" src="https://github.com/user-attachments/assets/fe288e6c-dfcb-49c9-a6e5-47c1d9c09bcd" />

### 프론트엔드 뉴스기사 출처 조회
<img width="2152" height="1658" alt="image" src="https://github.com/user-attachments/assets/0077bc2b-b69b-49d4-a45e-4b02d6d131ec" />

### 뉴스 기사 단건 조회
<img width="2242" height="1392" alt="image" src="https://github.com/user-attachments/assets/ca00782b-d5d1-4ea9-931f-f546f98ee970" />

### 전체 기사 보러가기
<img width="2566" height="1570" alt="image" src="https://github.com/user-attachments/assets/f2e296f6-ca13-4251-86a8-6d13ad47de04" />

### 조회수 중복 방지 기능 구현(Redis)
<img width="2366" height="1374" alt="image" src="https://github.com/user-attachments/assets/8486c028-cb21-420e-9810-3ada10a22e73" />

### 댓글 좋아요 종합 테스트
<img width="1666" height="1030" alt="image" src="https://github.com/user-attachments/assets/1ddb6388-067d-4f98-bba2-c660f101cee9" />


## 📚 Reference

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## 릴리스 노트

* **새로운 기능**
  * 뉴스 기사 검색 및 필터링 기능 추가 (키워드, 관심사, 출처별)
  * 기사 목록 조회 API 추가 (정렬 및 커서 기반 페이지네이션 지원)
  * 개별 기사 상세 조회 API 추가 (조회수 자동 집계)
  * 24시간 내 사용자별 중복 조회 방지 기능
  * 한글 콘텐츠 필터링으로 품질 개선
  * 기사 출처 목록 조회 API 추가

* **오류 처리**
  * 중앙화된 예외 처리기 추가로 일관된 오류 응답 제공
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yuhandemian committed:
fix: 기사 목록 API 커서 페이지네이션 오류 수정

- nextCursor를 UUID 형식으로 수정하여 무한 스크롤 문제 해결
- nextAfter를 createdAt 기반으로 변경하여 정확한 페이지네이션 구현
- Tuple 방식으로 쿼리 변경하여 createdAt 필드 접근 가능하도록 개선

yuhandemian committed:
feat: 전역 예외 헨들러 추가

yuhandemian committed:
fix: 코드래빗 피드백 반영

yuhandemian commented on monew/src/main/java/com/spring/monew/article/repository/impl/ArticleRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,239 @@
+          .exists();
+    } else {
+      viewedByMeExpression = Expressions.asBoolean(false);
+    }
```
@coderabbitai 나중에 기사 뉴스 단건 조회 기능 개발 할 때 처리할 예정 Redis로 조회수 제한을 걸 예정이고 기간은 24시간 으로 설정할 예정임

yuhandemian committed:
fix: ArticleSource enum 유효성 검증 추가

Namsoo315 commented on monew/src/main/java/com/spring/monew/article/controller/ArticleController.java:
```diff
@@ -0,0 +1,90 @@
+        }
+      }
+    }
+
```
위 위쪽이 이제 인가된 사용자만 처리하는거지요? 맞다면 추후에 인가된 사용자만 허용되게 하는 모든 Controller 단에서 처리하는 걸 이런식으로 하면 되지않을 가 싶습니다.

yuhandemian committed:
feat: 기사 춞처 목록 조회 기능 구현

yuhandemian committed:
feat: 기사 출처 목록 조회 기능 구현

haram-jo reviewed (approved):
수고하셨습니다!

yuhandemian committed:
fix: 들여쓰기 조정

yuhandemian committed:
feat: 뉴스 수집 데이터 품질 개선

- 한글 콘텐츠 필터링 추가 (NaverNewsApiClient, RssFeedClient)
  - containsKorean() 메서드로 제목/요약에 한글 포함 여부 확인
  - 한글이 없는 기사는 수집에서 제외

- 네이버 API 응답 요약 정제 로직 추가 (NaverNewsApiClient)
  - 영문 번역 안내문 제거 ("It assumes that there may be errors...")
  - URL 제거 (https://...)
  - 출처 표시 제거 ((출처1), (출처6))
  - 특수 기호 제거 (☞)
  - URL 인코딩 문자 제거 (%EC%82%AC...)
  - 한글이 포함된 라인만 추출
  - 요약 길이 제한 (500자)

이를 통해 RSS 피드(한경, 조선, 연합)는 깨끗한 긴 요약(300-500자),
Naver API는 짧지만 깨끗한 요약(150-200자)을 제공합니다.

yuhandemian committed:
feat: 뉴스 기사 상세 조회 API 구현 (#007)

GET /api/articles/{articleId} 엔드포인트 추가

주요 기능:
- 조회수 자동 증가 (Article.incrementViewCount())
- Redis 기반 24시간 중복 조회 방지
  - RedisConfig 추가 (RedisTemplate 설정)
  - "article:view:{articleId}:{userId}" 키로 중복 체크
- viewedByMe 필드 계산 (최근 24시간 내 조회 여부)
- ArticleView 엔티티에 조회 기록 저장
  - ArticleViewRepository 추가
  - ArticleView.of() 팩토리 메서드 추가

예외 처리:
- ArticleNotFoundException 추가 (404 반환)
- GlobalExceptionHandler에 핸들러 추가
- ErrorResponse 추가 (공통 에러 응답 구조)
- 잘못된 UUID 형식 시 400 반환

구현 파일:
- ArticleController.articleDetails()
- ArticleServiceImpl.getArticle(), trackView()
- ArticleDto 응답 (기존 활용)

yuhandemian committed:
refactor: userId 추출 로직 중복 제거

- extractUserId() private 헬퍼 메서드 추가
- articleList()와 articleDetails()에서 중복된 Principal 파싱 로직 제거
- 두 엔드포인트 모두 일관된 에러 처리 적용
  - 잘못된 UUID 형식 시 null 반환 (게스트 허용)
  - DRY 원칙 준수로 유지보수성 향상

코드 리뷰 피드백 반영

yuhandemian committed:
refactor: config 및 exception 파일을 common 패키지로 이동

- RedisConfig: config/ -> common/config/
- ErrorResponse: exception/ -> common/exception/
- 프로젝트 구조 일관성 개선

yuhandemian committed:
refactor: 미사용 ErrorResponse 파일 삭제

- ErrorResponse는 코드베이스에서 사용되지 않음
- GlobalExceptionHandler가 Spring의 ProblemDetail(RFC 7807)을 사용 중
- 중복 코드 제거로 에러 응답 처리 표준화

코드 리뷰 피드백 반영

Namsoo315 reviewed (approved):
문제없습니다. 고생하셨습니다.
